### PR TITLE
HttpSessionManager: add global_locked pool option.

### PR DESCRIPTION
**Pull request by traeak**
Use only when transaction loads starve the session pool to the point that the pool is mostly bypassed.
I'm not sure this is that useful unless someone has severe congestion issues.  Performance is not negatively impacted if the global locked isn't used.

NOTE: the critical section of the session pool was expanded due to the PR associated with this issue:  https://github.com/apache/trafficserver/issues/7096

### CODE REVIEW

File: doc/admin-guide/files/records.yaml.en.rst [hunk 1]
```diff
@@ -1067,24 +1067,36 @@ mptcp
...
+
+   For a hybrid pool, the operation starts as the global pool, but sessons
+   are returned to the local thread pool if the global pool lock is not
+   acquired rather than just closing the origin connection as is the
+   case in standard global mode.
+
+   For a ``global_locked`` pool connections are managed by a blocking
+   mutex instead of the normal try mutex.  Under extreme transaction
+   loads the connection pool starvation may result in most transactions
+   bypassing the connection pool resulting in runaway upstream
```
[the remark this answers is not in the document]
Comment (traeak) [reply doc/admin-guide/files/records.yaml.en.rst:1097]: retrying the lock would result in even further pool starvation.

**Comment by traeak** [comment]
[approve ci rocky]

File: doc/admin-guide/files/records.yaml.en.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (traeak) [reply doc/admin-guide/files/records.yaml.en.rst:1081]: retry on that lock would just cause congestion to go critical that much faster.

File: proxy/http/HttpSessionManager.cc [hunk 2]
```diff
@@ -146,52 +146,75 @@ ServerSessionPool::acquireSession(sockaddr const *addr, CryptoHash const &hostna
...
+    // This is broken out because only in this case do we check
+    // the host hash first. The range must be checked to verify
+    // an upstream that matches port and SNI name is selected.
+    // Walk backwards to select oldest.
+    // Prefer a session that's already on this thread.
+    in_port_t const port = ats_ip_port_cast(addr);
+    auto iter            = m_fqdn_pool.find(hostname_hash);
+    while (iter != m_fqdn_pool.end() && iter->hostname_hash == hostname_hash) {
+      Debug("http_ss", "Compare port 0x%x against 0x%x", port, ats_ip_port_cast(iter->get_remote_addr()));
+      auto *const netvc = dynamic_cast<UnixNetVConnection *>(iter->get_netvc());
```
Comment (masaori335) [inline proxy/http/HttpSessionManager.cc:164]: Do we need to compare `NetHandler`? Comparing thread like below (avoiding dynamic_cast) is enough? 
```
NetVConnection *const netvc = iter->get_netvc();
...
  if (netvc->thread == ethread) {
```

Comment (traeak) [reply proxy/http/HttpSessionManager.cc]: got it thanks.

File: proxy/http/HttpSessionManager.cc [hunk 3]
```diff
@@ -146,52 +146,75 @@ ServerSessionPool::acquireSession(sockaddr const *addr, CryptoHash const &hostna
...
-          zret = HSM_DONE;
-          break;
+      while (iter != m_ip_pool.end() && ats_ip_addr_port_eq(iter->get_remote_addr(), addr)) {
+        auto *const netvc = iter->get_netvc();
+        if ((!(match_style & TS_SERVER_SESSION_SHARING_MATCH_MASK_HOSTONLY) || iter->hostname_hash == hostname_hash) &&
+            (!(match_style & TS_SERVER_SESSION_SHARING_MATCH_MASK_SNI) || validate_sni(sm, netvc)) &&
+            (!(match_style & TS_SERVER_SESSION_SHARING_MATCH_MASK_HOSTSNISYNC) || validate_host_sni(sm, netvc)) &&
+            (!(match_style & TS_SERVER_SESSION_SHARING_MATCH_MASK_CERT) || validate_cert(sm, netvc))) {
+          // Check if on same thread
+          if (netvc->thread == ethread) {
```
Comment (shinrich) [inline proxy/http/HttpSessionManager.cc:203]: How is this only triggered in the global locked case?

**Comment by shinrich** [comment]
I like the idea of this.  One question in the code.

How did this compare to hybrid?  Fewer connections to origin opened?

Did you see performance degradation due to walking longer session lists in the global_locked vs global case?

File: proxy/http/HttpSessionManager.cc (hunk 3 above)
Comment (traeak) [reply proxy/http/HttpSessionManager.cc:203]: it's not this one is actually a change to the pools in general.  if necessary i could split this into 2 prs.

**Comment by traeak** [comment]
I'm going to run performance tests on one of our heavily loaded linear only
caches in the near future to see how it performs.  I still need to check
the performance difference since removing the dynamic cast for each list
item.

File: proxy/http/HttpSessionManager.cc (hunk 3 above)
Comment (traeak) [reply proxy/http/HttpSessionManager.cc:203]: done and tested.

**Comment by traeak** [comment]
I ran some performance tests and the session pool walking has a negative affect if there's a large number of matching pool entries.  I will cut that number back to only checking the first few (whatever that number is, 4 or 8) for same thread session, otherwise fall back to taking the first (most idle) session.

**Comment by traeak** [comment]
> I like the idea of this. One question in the code.
> 
> How did this compare to hybrid? Fewer connections to origin opened?
> 
> Did you see performance degradation due to walking longer session lists in the global_locked vs global case?

Walking the list can hurt performance.  I will test using a walk threshold and generate a new PR if it's worth it.

**Review by masaori335 — APPROVED** [review]
Having an option of  blocking mutex instead of try lock looks reasonable because it's faster in some cases. Let's land this and add `prefer session on same thread` feature later.

**Pull request merged**